proposals/0010-validating-resource-container-elements.md

```diff
@@ -0,0 +1,142 @@
+* Proposal: [0010](0010-validating-resource-container-elements.md)
+* Author(s): [Joshua Batista](https://github.com/bob80905)
+* Sponsor: TBD
+* Status: **Under Consideration**
+* Impacted Project(s): (LLVM)
+
+*During the review process, add the following fields as needed:*
+
+* PRs: [#NNNN](https://github.com/microsoft/DirectXShaderCompiler/pull/NNNN)
+* Issues: [#75676](https://github.com/llvm/llvm-project/issues/75676)
+
+## Introduction
+Resources are often used in HLSL, with various resource element types (RETs).
+For example:
+```
+RWBuffer<float> rwbuf: register(u0);
+```
+In this code, the RET is `float`, and the resource type is `RWBuffer`. The
+resource type is not a `RawBuffer` variant, and so there is a distinct set
+of rules that define valid RETs for this resource type.
+
+RETs for non-`RawBuffer` variants may include basic types (ints and uints of sizes 16
+and 32, as well as half, and float). Structs that contain fields of these basic

```

Comment:
bool, uint64_t and double are currently allowed as well https://godbolt.org/z/eb7Yh73W6

Reply:
If I understand correctly from my conversation with @tex3d, though the `uint64_t` and `double` cases compile, they will cause undefined behavior at runtime, and we do not want to allow any types that are larger than 32 bits.
Even if 2 64-bit types can technically fit in 4 32-bit fields, it's a hard limit that there may not be more than 4 elements, and each element may not be more than 32 bits.
However, I'm unsure if we can allow `bool`, I think we can.

Reply:
Support for 64-bit values seems like a totally reasonable thing to do.  Is there something fundamental about them that mean we can't support them properly, or is this just that DXC is buggy?

What about in SPIR-V today?  Looks like SPIR-V supports `int64_t2`.

Reply:
It's something fundamental (If I recall correctly, I think it had to do something with the hardware converter / DXGI format being capped at 32 bits in D3D, and that is not changeable).

Reply:
I don't think I said anything about them not working or causing undefined behavior.  The undefined behavior case was tied to matching unorm or snorm in HLSL with the DXGI resource type for typed UAVs.

They definitely are supported on DXC (by decomposing to 32-bit components), and FXC the same way IIRC (for double).

Reply:
Perhaps what was thought of was the fact that you can't use these with Sample operations (only Load/Store), since it's equivalent to a typed resource with an int32 component type, which doesn't support those operations.

In fact, if you try it, you get a validation error (we only catch it there), as it attempts to use a 64-bit overload for the sample operation, and define an invalid `%dx.types.ResRet.f64` type.

Reply:
> since it's equivalent to a typed resource with an int32 component type, which doesn't support those operations.

Circling back to this conversation, I recall that this isn't correct anymore - there is more to add.

We added [integer sampling support in SM 6.7](https://microsoft.github.io/DirectX-Specs/d3d/HLSL_SM_6_7_Advanced_Texture_Ops.html#integer-sampling).  This relaxed checks in the front end that disallowed integer types.  So integer types are actually allowed in SM 6.7 on (but not for comparison sampling).

Unfortunately, when 64-bit types are used with sample operations, those resource return types won't be translated into two i32 values per 64-bit value.  That's what leads to the invalid operation type for these caught by validation.  These types could be supported using integer sampling, if we had broken them up as we do for load operations.

I think this means that when using a 64-bit component type in a typed resource element, we should consider this as being implicitly translated to two 32-bit uints per 64-bit component for the actual type used.  Then the rules are applied according to the translated element type.

There is one thing bugging me: double component type seemingly *should* trigger an error in `IsValidObjectElement`, but it doesn't.  See the code here:
https://github.com/microsoft/DirectXShaderCompiler/blob/a023a95f73618e8ca1791147d3722a3e1ea7608f/tools/clang/lib/Sema/SemaHLSL.cpp#L6058